JSON 문자열 분석기 2단계 - Object 분석

popsmile opened the pull request:
아래의 내용을 추가하여 2단계 Object 분석 요구사항을 구현해보았습니다.
1. `enum JSONValue` 를 만들어 JSON 데이터 구조의  VALUE 부분에 담길 수 있는 타입으로만 구성했습니다.
2. `JSONDataForm` 프로토콜을 만들어, 새롭게 추가한 `JSONObject` 와 `JSONArray` 구조체가 이 프로토콜을 따르도록 설계했습니다.
3. `JSONObject` 와 `JSONArray` 두 구조체의 객체가 타입 개수를 카운트할 시 `TypeInspector` 를 통해서 객체 스스로 일하도록 했습니다.
4. `JSONParser` 가 문자열을 감싸고 있는 괄호에 따라 `JSONObject` 또는 `JSONArray` 를 생성하도록 했습니다. (어려움을 겪고 있는 점: `JSONParser` 내부의 `makeJSONArray()` 메소드를 분리하고자 여러 시도를 해보았으나 실패하였습니다..ㅠ 해당 메소드 로직은 조금 더 생각해보겠습니다. 정규식을 공부해서 적용해볼까 생각 중 입니다. )

popsmile committed:
Rename JSONGenerator to JSONParser

Because the functions in JSONGenerator is not generate JSON data but parse them to Swift data structures.

popsmile committed:
Create an enum JSONValue

The enum JSONValue only includes the types allowed to be 'value' of JSON data.

popsmile committed:
Add extensions of String

popsmile committed:
Add makeJSONObject method

Depending on the prefix of the parameter String, JSONObject or JSONArray is made by parsing the string.

popsmile committed:
Create a protocol JSONDataForm and two structs JSONObject, JSONArray

The structs JSONObject and JSONArray conform to the protocol JSONDataForm

popsmile committed:
Revise how to count types through OutputView

popsmile committed:
Separate structs JSONObject and JSONArray to each file

popsmile committed:
Move protocol JSONDataForm to a separated file

popsmile committed:
Divide makeJSONObject method

popsmile committed:
Fix the wrong logic of makeJSONArray method

popsmile committed:
Simplify makeJSONArray method

popsmile committed:
Reformat the parameter names of makeSentence method

popsmile committed:
Update unit tests according to the changes of codes

godrm commented on JSONParser/JSONParser/TypeInspector.swift:
```diff
@@ -12,22 +12,26 @@ public enum typeName: String {
+    case object = "객체"
+    case array = "배열"
     case none = ""
 }
```
enum 이더라도 타입으로 쓰는 경우는 대문자로 시작하는게 규칙입니다. 
JSONValue 와 형태가 비슷한데 새로운 타입을 만들게 아니라 기능으로 합칠 방법은 없을까요?

popsmile replied:
네. `enum JSONValue` 의 `var typeName` 을 추가하여 기능을 합쳤습니다.

godrm commented on JSONParser/JSONParser/JSONValue.swift:
```diff
@@ -0,0 +1,17 @@
+    case int(Int)
+    case bool(Bool)
+    case object([String : JSONValue])
+    case array([JSONValue])
```
JSONArray와 JSONObject 타입을 만들었는데 JSONValue 에서는 사용하지 않네요. 
그대로 사용해도 되지 않을까요?

popsmile replied:
넵. 만든 타입으로 사용하도록 수정했습니다.

godrm commented on JSONParser/JSONParser/OutputView.swift:
```diff
@@ -16,15 +16,15 @@ struct OutputView {
-        var typeNames: [String?] = [typeName.string.rawValue, typeName.int.rawValue, typeName.bool.rawValue]
-        let typeCount = TypeInspector.countType(of: jsonArray)
+    private static func addTypeName(to jsonDataForm: JSONDataForm) -> String {
+        var typeNames: [String?] = [typeName.string.rawValue, typeName.int.rawValue, typeName.bool.rawValue, typeName.object.rawValue]
```
typeNames는 항상 같은 값을 갖고 있으니까 상수로 혹은 static 객체로 만들어도 되지 않을까요?

popsmile replied:
타입 개수가 0개 일 때, `typeNames` 에서 해당 타입 이름을 `nil` 로 변경하기 위해서 변수로 두었습니다. 지금은 타입 이름을 출력하는 로직을 수정하여,  필요하지 않아 삭제하였습니다.

godrm commented on JSONParser/JSONParser/TypeInspector.swift:
```diff
@@ -12,22 +12,26 @@ public enum typeName: String {
 
 struct TypeInspector {
-    static func countType(of jsonArray: [Any?]) -> [String: Int] {
+    static func countType(of jsonValues: [JSONValue]) -> [String: Int] {
```
`countType(of jsonValues: [JSONValue])` 을 배열만 만든 이유가 뭔가요?
이러면 JSONObject는 전체 dictionary 값들을 배열로 만들어야 하겠죠 (이렇게 `TypeInspector.countType(of: Array(jsonObject.values))`)
이러면 Dictionary -> Array로 바꾸는 비용이 발생합니다. JSON 데이터가 커질수록 부담이 되고 느려지는 방식입니다.
차라리 데이터 구조를 바꾸기 보다는 그대로 탐색하는 메소드를 하나 더 만드세요.

popsmile replied:
아하 그 비용에 대해서는 생각지못했습니다 ㅠㅠ 참고하겠습니다. 
`TypeInspector` 의 `countType` 파라미터가 `[JSONValue]` 이었기 때문에, Dictionary 의 value 값만 전해주기 위해 Array 로 바꿔주었습니다. 
지금은 `TypeInspector` 구조체를 삭제하고 `JSONObject` 와 `JSONArray` 구조체 내부에서 `countType` 을 각각의 로직으로 수행하도록 수정했습니다. `JSONObject` 의 `countType()`의 경우, dictionary.values 를 사용하여 해당 컬렉션을 그대로 탐색하도록 하였습니다.

godrm commented on JSONParser/JSONParser/JSONParser.swift:
```diff
@@ -0,0 +1,83 @@
+                guard let jsonValue = typeCast(from: slice.trimWhiteSpaces()) else { continue }
+                jsonArray.append(jsonValue)
+            }
+            iterateIndex = jsonString.index(after: iterateIndex)
```
와.. 이 메소드 복잡하네요. 개선할 방법은 없을까요?

popsmile replied:
네.. 계속해서 고민하다가, 정규표현식을 사용해서 `JSONArray` 속  `JSONObject` 를 추출하는 방식으로 개선해보았습니다.

godrm reviewed (commented):
점점 구현이 복잡해지고 있습니다. 
어떤 부분을 개선할 지 다음 피드백을 확인해보세요.

popsmile committed:
Simplify makeJSONObject() and makeJSONArray() with regular expression

popsmile committed:
Cut enum typeName and add it to enum JSONValue

Add var typeName: String in enum JSONValue, which was in enum typeName. In JSONObject and JSONArray, countType methods are also simplified to cut the struct TypeInspector

popsmile committed:
Cut TypeInspector

popsmile committed:
Fix init() parameter of JSONObject and JSONArray to _

popsmile committed:
Include structs JSONObject and JSONArray as cases of enum JSONValue

popsmile committed:
Simplify addTypeName()

popsmile committed:
Update unit tests

Cut UnitTestTypeInspector

popsmile committed:
Create and Update unit tests

popsmile commented:
피드백주신 내용을 바탕으로 수정해보았습니다. 
- 만들어놓은 `JSONObject` , `JSONArray` 구조체를 `enum JSONValue` 에서 타입으로 사용했습니다.
- `makeJSONArray()` 메소드를 개선해보았습니다. 
- `TypeInspector` 를 삭제하고, 이 구조체를 통해 수행했던 `countType()` 을 `JSONObject` , `JSONArray` 구조체 내부에서 각각 수행하도록 수정했습니다.

popsmile committed:
Fix issues of codacy

popsmile committed:
Fix issues of codacy

popsmile committed:
Fix an issue of codacy
